src/cdomains/intDomain.ml

```diff
@@ -837,12 +838,8 @@ struct
     | `Excluded (_, xr), `Excluded(_, yr) ->
       check_identical_range xr yr;
       `Excluded(S.empty (), xr)
-    (* If any one of them is bottom, we return top *)
-    | `Bot, `Excluded(_, xr)
-    | `Bot, `Definite(_, xr)
-    | `Excluded(_ , xr), `Bot
-    | `Definite(_, xr), `Bot -> top ()
-    | `Bot, `Bot -> top ()
+    (* If any one of them is bottom, we return bottom *)
+    | _ -> `Bot
```

Comment:
This may get rid of the different ikind problem, but is probably unsound.
How does it happen? There should be no operations on `Bot at all, no? And if there were it should still yield `Top.

Reply:
It is also what we do in `Interval32` e.g.
https://github.com/goblint/analyzer/blob/0bcce88f7302c6a45d1e85859e65a8626ebe1dba/src/cdomains/intDomain.ml#L268-L272

and for functions that are injective in eitherargument (i.e. `lift2_inj`)

https://github.com/goblint/analyzer/blob/0bcce88f7302c6a45d1e85859e65a8626ebe1dba/src/cdomains/intDomain.ml#L775-L800

The place where this happens is in the `ptx_comb.c` example in coreutils: https://github.com/goblint/bench/blob/master/coreutils/ptx_comb.c (Line 69313 / ptx.c:948)

```
if ((0xffffffffffffffffUL / sizeof(*(occurs_table[0])) - 1UL) / 2UL < occurs_alloc[0]) {
```
It happens during the call to the new `invariant`:
https://github.com/goblint/analyzer/blob/406b193a4d2900fdebc785c37c348fd1278c6fb4/src/analyses/base.ml#L1409
https://github.com/goblint/analyzer/blob/406b193a4d2900fdebc785c37c348fd1278c6fb4/src/analyses/base.ml#L1404



**Could you please explain again why it is unsound to propagate `Bot here? I don't think I understand this yet.**

Reply:
https://github.com/goblint/analyzer/blob/406b193a4d2900fdebc785c37c348fd1278c6fb4/src/analyses/base.ml#L847-L852

I understand why we introduced this in the base analysis at some-point to be sound for programs with UB, but I don't get why we need to do it inside all domains as well.

Reply:
The whole bot/top handling is pretty brittle.
Is any of the values from a global invariant?
My point was about local state - there no variable should ever be \`Bot. But maybe it happens if something is uninitialized. Then it would be unsound to go to \`Bot, e.g. `uninit + 1` should be \`Top.
Global variables are initialized with zero, so there also should be no \`Bot for global invariants.
I guess for `malloc` we start out with \`Bot. Still, same situation.
The idea of the old code as in `Interval32` probably was to ignore all accesses to uninitialized variables since it is undefined behavior. Then, going to \`Bot makes the whole analysis faster and more precise, but only sound if no such accesses happen.
However, I thought that's what we wanted to improve on. Otherwise the results are pretty useless since no one knows if their program is free of undefined behavior.

Reply:
But are precisely those cases not the ones that are handled by the snippet above already?
This `bot` would bubble up until it arrives at this `eval_rv` call where it then is replaced with `top`. Or is there something I am missing here?

Reply:
Ah, replied before seeing the `eval_rv` comment.
Maybe it's all fine then.
In retrospect this seems hacky. From a domain view the result should be \`Top. Why didn't we do this?

Reply:
Helmut and I were both under the impression, that from a domain perspective, these functions should be strict. So the behavior with returning bottom if one of the arguments is bottom would be the right thing to do.

Reply:
`unop bot = bot` and `binop bot bot = bot` are strict, I don't know what you would call this.
The problem is that it becomes unsound if we forget to change from bot to top some place and there's no way to notice.
It's safer to raise an exception and catch it in `eval_rv`. Result should be the same.

Reply:
I know raise an exception whenever we compute a binop and one of the arguments is `Bot`. 
But I'm not too happy with it, doesn't really seem clean either.

Right now I changed it only for `Interval32` and `Def_Exc`, if we decide to do this, I will also change it for the other domains.

Reply:
Btw, raising instead of going to bottom seems like this was the right decision engineering-wise. It helped me spot a quite severe bug in my WCET fork.